src/software/embedded/motor_controller/stspin_motor_controller.cpp

```diff
@@ -0,0 +1,239 @@
+
+    // SET_SPEEDRAMP expects the ramp time in millis to be in register bx.
+    // We do speed ramping ourselves in MotorService, so we just want to
+    // set the target speed without ramping (hence we set reg bx to 0).
+    sendAndReceiveFrame(motor, StSpinOpcode::MOV_BX, 0);
+
+    sendAndReceiveFrame(motor, StSpinOpcode::SET_SPEEDRAMP);
+
+    return current_velocity;
+}
+
+void StSpinMotorController::immediatelyDisable()
+{
+    driver_control_enable_gpio_->setValue(GpioState::LOW);
+}
+
+void StSpinMotorController::openSpiFileDescriptor(const MotorIndex& motor)
+{
+    file_descriptors_[CHIP_SELECTS.at(motor)] = open(SPI_PATHS.at(motor), O_RDWR);
+    CHECK(file_descriptors_[CHIP_SELECTS.at(motor)] >= 0)
+        << "can't open device: " << motor << "error: " << strerror(errno);
+
+    int ret =
+        ioctl(file_descriptors_[CHIP_SELECTS.at(motor)], SPI_IOC_WR_MODE32, &SPI_MODE);
+    CHECK(ret != -1) << "can't set spi mode for: " << motor
+                     << "error: " << strerror(errno);
+
+    ret = ioctl(file_descriptors_[CHIP_SELECTS.at(motor)], SPI_IOC_WR_BITS_PER_WORD,
+                &SPI_BITS);
+    CHECK(ret != -1) << "can't set bits_per_word for: " << motor
+                     << "error: " << strerror(errno);
+
+    ret = ioctl(file_descriptors_[CHIP_SELECTS.at(motor)], SPI_IOC_WR_MAX_SPEED_HZ,
+                &MAX_SPI_SPEED_HZ);
+    CHECK(ret != -1) << "can't set spi max speed hz for: " << motor
+                     << "error: " << strerror(errno);
+}
+
+int16_t StSpinMotorController::sendAndReceiveFrame(const MotorIndex& motor,
+                                                   const StSpinOpcode opcode,
+                                                   const int16_t data)
+{
+    //  A frame is 6 bytes long and has the following format:
+    //
+    //  +-------+--------+---------------+-------+-------+
+    //  |  SOF  | OPCODE |     DATA      |  CRC  |  EOF  |
+    //  +-------+--------+---------------+-------+-------+
+    //      0       1        2       3       4       5
+    //
+    //  For slave frames, OPCODE will be either an ACK or NACK acknowledging
+    //  receival of the last master frame.
+    //
+    //  The byte order of DATA is big endian; therefore the MSB is transmitted first.
+    //  DATA may be ignored if the operation has no data to transmit/receive.
+
+    uint8_t tx[FRAME_LEN] = {0};
+    uint8_t rx[FRAME_LEN] = {0};
+
+    tx[0] = FRAME_SOF;
+    tx[1] = static_cast<uint8_t>(opcode);
+    tx[2] = static_cast<uint8_t>(0xFF & (data >> 8));
+    tx[3] = static_cast<uint8_t>(0xFF & data);
+    tx[5] = FRAME_EOF;
+
+    const uint8_t message[] = {tx[1], tx[2], tx[3]};
+    tx[4]                   = Crc8Autosar::calc(message, sizeof(message));
+
+    spiTransfer(file_descriptors_[CHIP_SELECTS.at(motor)], tx, rx, FRAME_LEN,
+                SPI_SPEED_HZ);
+
+    return static_cast<int16_t>((static_cast<uint16_t>(rx[2]) << 8) | rx[3]);
```

Comment:
What's in `rx[0]` and `rx[1]`

Reply:
`rx[0]` has the start of frame marker and `rx[1]` has an ACK or NACK code indicating whether the last frame was properly received. This is documented in the comment block above